test/unit/nested.cpp

```diff
@@ -0,0 +1,31 @@
+#include "gtest/gtest.h"
+
+#include "RAJA/pattern/nested.hpp"
+
+#include <iostream>
+
+TEST(Nested, Basic)
+{
+  using namespace RAJA::nested;
+  using Index_type = RAJA::Index_type;
+  using pol = Policy<For<1, RAJA::seq_exec>, For<0, RAJA::seq_exec>>;
+  RAJA::nested::forall(pol{},
+                       RAJA::util::make_tuple(RAJA::RangeSegment(0, 5),
+                                              RAJA::RangeSegment(0, 5)),
+                       [=](Index_type i, Index_type j) {
+                         std::cout << i << " " << j << std::endl;
+                       });
+}
+
+TEST(Nested, collapse)
+{
+  using namespace RAJA::nested;
+  using Index_type = RAJA::Index_type;
+  using pol = Policy<Collapse<RAJA::seq_exec, CFor<1>, CFor<0>>>;
```

Comment:
@rhornung67 : we have a contributor's code now, you're not allowed to impugn my excellent credentials at playing the fool.

Somewhat more seriously, I do think before this is merged we should get a lot of us who don't use forallN to try to write code using this, and document the experience. I think forallN is great but a little scary, and some "here's how to get started with it" docs could be good, wind up in the tutorial, etc.

That's outside the scope of the review, but maybe @trws when you think there's enough to play with (tiling, CUDA) ping me and I'll test it.